lib/hashie/extensions/coercion/active_model.rb

```diff
@@ -0,0 +1,85 @@
+require 'active_model/type'
+
+module Hashie
+  module Extensions
+    module Coercion
+      module ActiveModel
+        # Symbol is the one glaring type omission. Define it if it is not
+        # already defined.
+        begin
+          ::ActiveModel::Type.lookup(:symbol)
+        rescue ArgumentError
+          class Symbol < ::ActiveModel::Type::String
+            def type
+              :symbol
+            end
+
+            private
+
+            # The symbol type leverages in coercion logic in the String
+            # superclass and simply uses `to_sym` to covert to a symbol.
+            def cast_value(value)
+              super.to_sym
+            end
+          end
+          ::ActiveModel::Type.register(:symbol, Symbol)
+        end
+
+        # Array of symbols of the included ActiveModel types.
+        ACTIVE_MODEL_TYPES = ::ActiveModel::Type.registry
+                             .send(:registrations)
+                             .map { |r| r.send(:name) }
+                             .freeze
+
+        def build_coercion(type)
+          if type.is_a? Enumerable
+            if type.class <= ::Hash
+              type, key_type, value_type = type.class, *type.first
+              build_hash_coercion(type, key_type, value_type)
+            else # Enumerable but not Hash: Array, Set
+              value_type = type.first
+              type = type.class
+              build_container_coercion(type, value_type)
+            end
+          elsif ACTIVE_MODEL_TYPES.include? type
+            build_active_model_type_coercion(type)
+          elsif type.respond_to? :coerce
+            lambda do |value|
+              return value if value.is_a? type
+              type.coerce(value)
+            end
+          elsif type.respond_to? :new
+            lambda do |value|
+              return value if value.is_a? type
+              type.new(value)
+            end
+          else
+            fail TypeError, "#{type} is not a coercable type"
+          end
+        end
+
+        def build_hash_coercion(type, key_type, value_type)
+          key_coerce = fetch_coercion(key_type)
```

Comment:
We're entering message-tracing hell here. Is `#fetch_coercion` still from the `Hashie::Extensions::Coercion` module? It's hard to tell and will be hard to debug if there's a bug here.

Can we think of a better way to structure this so it's more apparent?